velox/functions/lib/QuantileDigest.h

```diff
@@ -0,0 +1,850 @@
+        }
+      },
+      rights_,
+      lefts_);
+  return std::min(postprocessByType(max_), result);
+}
+
+template <typename T>
+QuantileDigest<T>::U QuantileDigest<T>::lowerBound(int32_t node) const {
+  if constexpr (std::is_same_v<U, int64_t>) {
+    uint64_t mask = 0L;
+    if (levels_[node] > 0) {
+      mask = static_cast<uint64_t>(-1L) >> (64 - levels_[node]);
+    }
+    return bitsToLong(values_[node] & ~mask);
+  } else {
+    uint32_t mask = 0;
+    if (levels_[node] > 0) {
+      mask = static_cast<uint32_t>(-1L) >> (32 - levels_[node]);
+    }
+    return bitsToLong(values_[node] & ~mask);
+  }
+}
+
+template <typename T>
+QuantileDigest<T>::U QuantileDigest<T>::upperBound(int32_t node) const {
+  if constexpr (std::is_same_v<U, int64_t>) {
+    uint64_t mask = 0L;
+    if (levels_[node] > 0) {
+      mask = static_cast<uint64_t>(-1L) >> (64 - levels_[node]);
+    }
+    return bitsToLong(values_[node] | mask);
+  } else {
+    uint32_t mask = 0L;
+    if (levels_[node] > 0) {
+      mask = static_cast<uint32_t>(-1) >> (32 - levels_[node]);
+    }
+    return bitsToLong(values_[node] | mask);
+  }
+}
+
+template <typename T>
+int64_t QuantileDigest<T>::estimatedInMemorySizeInBytes() const {
+  auto nodeCount = nextNode_ - freeCount_;
+  return /*version*/ sizeof(char) + sizeof(maxError_) +
+      /*alpha*/ sizeof(double) + /*landmarkInSeconds*/ sizeof(int64_t) +
+      /*min*/ sizeof(int64_t) + /*max*/ sizeof(int64_t) +
+      /*nodeCount*/ sizeof(int32_t) +
+      (sizeof(typename decltype(counts_)::value_type) * nodeCount) +
+      (sizeof(typename decltype(levels_)::value_type) * nodeCount) +
+      /*values*/ (sizeof(int64_t) * nodeCount);
+}
+
+template <typename T>
+void write(T value, char*& out) {
+  folly::storeUnaligned(out, value);
+  out += sizeof(T);
+}
+
+template <typename T>
+int64_t QuantileDigest<T>::serialize(char* out) {
+  compress();
+  const char* outStart = out;
+  write<char>(0, out); // version
+  write<double>(maxError_, out);
+  write<double>(0.0, out); // alpha
+  write<int64_t>(0, out); // landmarkInSeconds
+  write<int64_t>(static_cast<int64_t>(min_), out);
+  write<int64_t>(static_cast<int64_t>(max_), out);
+  write<int32_t>(nextNode_ - freeCount_, out);
+  postOrderTraverse(
+      root_,
+      [&](int32_t node) {
+        if constexpr (std::is_same_v<U, int64_t>) {
+          auto nodeStructure =
+              static_cast<int8_t>(std::max(levels_[node] - 1, 0) << 2);
+          if (lefts_[node] != -1) {
+            nodeStructure = static_cast<int8_t>(nodeStructure | 1);
+          }
+          if (rights_[node] != -1) {
+            nodeStructure = static_cast<int8_t>(nodeStructure | 2);
+          }
+          write<int8_t>(nodeStructure, out);
+          write<double>(counts_[node], out);
+          write<int64_t>(values_[node], out);
+
+        } else {
+          auto nodeStructure = static_cast<int8_t>(
+              std::max((levels_[node] == 32 ? 64 : levels_[node]) - 1, 0) << 2);
```

Comment:
If levels_[node] == 32, it means the left and right children of node share 0 contiguous common bits from the most significant side, which means their sign bits are different. Since Java store the values always in int64_t, when the sign bit are different, levels_[node] would be 64 - 0 = 64. So we when we see levels_[node] == 32, we adjust it back to 64 to match Presto's serialization.

Reply:
This should only happen when `U = int32_t`, let's add that condition and extract this to a separate method